### ISSUE TITLE
Merge Fields Not Populating in Communication Wizard

### ISSUE THREAD

**Issue by MicahO** [SparkDevNetwork/Rock#4436]
### Description
When using the Email Wizard, merge fields under {{ Person }} do not populate using the drop-down. They work properly if typed by hand, but the drop-down won't populate them.

NOTE: This does not reproduce on the sandbox site as of 9/29/2020.

### Steps to Reproduce
Not reproducible on test site, but see this YouTube link for a demonstration.

https://youtu.be/VE0-y9EBDmo

**Expected behavior:**

Lava selected in the dropdown should populate in the email.

**Actual behavior:**

Lava selected in the dropdown is not populating the email.

### Versions
* **Rock Version:** [11.1]
* **Client Culture Setting:** [en-US]

**Comment by smross** [comment]
@MicahO I think I'm experiencing the same. Here's what my environments look like

TL;DR I am experiencing what may be _two_ bugs:

- "Second-level merge field picker doesn't insert merge fields"
- Merge field picker modal doesn't show your selected merge field in the modal (assuming modal is the correct word)

**Production Environment**

- Hosted in Azure (AzureSQL + IIS VM)
- Version 11.2
- Mailgun as transport
- Using Safari and Google Chrome as the browsers
- Using macOS and ChromeOS as the OS'
- _Typing_ merge fields works as expected
- _Selecting_ 'top-level' merge fields such as `Subject`, `From Email`, `From Name` in the merge field UI modal works, **however**, there is an unexpected UI experience where the selected merge field _does not_ appear in the modal. The "X" for removing the merge field selection does appear if you hover over where the modal should be. Selecting 'OK' does actually insert the merge fields
- _Selecting_ 'second-level' merge fields in the merge field UI modal **does not** work. You can select it in the modal, the unexpected UI experience where the selected merge field _does not_ appear in the modal. The "X" for removing the merge field selection does appear if you hover over where the modal should be. But selecting 'OK' doesn't actually insert it

Example video: [https://drive.google.com/file/d/1KNzDXkowBNmgP-m00pMU2zx--mXcur2b/view?usp=sharing](https://drive.google.com/file/d/1KNzDXkowBNmgP-m00pMU2zx--mXcur2b/view?usp=sharing)

**Sandbox Environment**

- Hosted in Azure (AzureSQL + IIS VM)
- Version 11.3 (installed from the Beta update source)
- Mailgun as transport (but in a sandbox/test limitation)
- Using Safari and Google Chrome as the browsers
- Using macOS and ChromeOS as the OS'
- _Typing_ merge fields works as expected
- _Selecting_ 'top-level' merge fields such as `Subject`, `From Email`, `From Name` in the merge field UI modal works, **however**, there is an unexpected UI experience where the selected merge field _does not_ appear in the modal. The "X" for removing the merge field selection does appear if you hover over where the modal should be. Selecting 'OK' does actually insert the merge fields
- _Selecting_ 'second-level' merge fields in the merge field UI modal **does not** work. You can select it in the modal, the unexpected UI experience where the selected merge field _does not_ appear in the modal. The "X" for removing the merge field selection does appear if you hover over where the modal should be. But selecting 'OK' doesn't actually insert it

**Comment by jennifertroeger** [comment]
We are on v12 and experiencing this as well (selecting from beneath the Person model). I tested on both Firefox and Chrome. Typing the lava in manually does work.

**Comment by smross** [comment]
Jennifer, although I'm sad it's also happening to your org, I'm thankful
you're also chiming in. Have you experienced this both on custom templates
and 'stock' templates? Is there anything about your environment that gives
you clues?

**Comment by jennifertroeger** [comment]
It's an issue on both the Default and custom templates. As you stated, if I select the top level category `{{ FromAddress }}` it works correctly. Anything nested beneath that doesn't work. If you flip over to the </> side you can see that nothing has been inserted.

**Comment by smross** [comment]
@tbunney thanks for confirming it's also happening for you!

I'll see if we can get some more momentum on what each of us experiencing this can do to provide diagnostics so we can determine why this is happening.

**Comment by leahjennings** [comment]
Also noticed this, when accessing the API tool bench (`<your_server_url>/api/docs/index#!/MergeFields/GETapi_MergeFields_id`), that whenever I try to do get a merge field that's level-two or deeper, the `Accept` header is `text/html`. When doing the same query on the demo's API tool bench, the `Accept` header is `application/json`.

Our instance:
<img width="981" alt="Screen Shot 2021-03-30 at 3 47 07 PM" src="https://user-images.githubusercontent.com/66277785/113047705-a3c62000-916f-11eb-88c0-57023fab3348.png">

Demo site:
<img width="999" alt="Screen Shot 2021-03-30 at 3 50 41 PM" src="https://user-images.githubusercontent.com/66277785/113047810-c6583900-916f-11eb-87a9-f7f053ef2920.png">

**Comment by leahjennings** [comment]
This one was really bothering me, so I kept digging. Turns out, there's a setting in our web.config that was causing this behavior (shoutout to StackOverflow: https://stackoverflow.com/questions/52968137/pipe-character-in-asp-net-route-parameter-causing-404).

We had this line:
`<httpRuntime targetFramework="4.5.2" requestValidationMode="4.5" requestValidationType="Rock.Web.RequestValidator" relaxedUrlToFileSystemMapping="true" maxRequestLength="102400" waitChangeNotification="5" maxQueryStringLength="16384"/>`

Testing in our development environment, I set `relaxedUrlToFileSystemMapping` to `"false"`, restarted IIS, and I get the correct results from the MergeFields API whenever I'm requesting a level-two or deeper.

My only question though, which I'm hoping some from the core team could answer, is are we messing up anything if we change `relaxedUrlToFileSystemMapping="true"` to false? I'm not an IIS guru, so not sure what else, if anything, that setting would change.

@smross can you check in your environment's web.config to see if you have the same setting as we do?

**Comment by azturner** [comment]
Not sure if it's related, but I actually had a comment on that StackOverflow question that for some reason is no longer visible. Posting it here in case it helps:
_This turned out to be due to an empty "NOT_A_VALID_FILESYSTEM_PATH" folder in the wwwroot folder. After deleting that folder, the route worked and no longer returned a 404 error. Not sure why. It does appear to be related to the relaxedUrltoFileSystemMapping setting (which was set to true), but not quite sure how._

**Comment by azturner** [comment]
A church we work with also reported this problem and I just checked their wwwroot folder and they had a file named "NOT_A_VALIDFILESYSTEM_PATH". As soon as I deleted that file, the merge fields starting working again.  Still not sure why or how that file gets created though.

**Comment by MicahO** [comment]
Holy cow @azturner that was it. Well done. And @leahjennings you are amazing. If you're going to Rx2021 this year, I owe you a beverage.

**Comment by leahjennings** [comment]
Deleting that file on our instance also fixed it for us, _without_ changing the web.config. Great catch @azturner! And yeah, @MicahO planning to be there. :)

**Comment by cabal95** [comment]
@azturner We (Shepherd) found that file was being created when certain files were uploaded to our Rock instance. Specifically certain characters in the filename. Which seems to correlate with this other SO article: https://stackoverflow.com/questions/26789048/asp-net-server-mappath-returns-the-string-not-a-valid-filesystem-path

I think it was when we still had one of our File Types set to store the files on disk rather than in the cloud, so if somebody tried to upload a file with a `:` in the filename (for example a screenshot from a Mac), it would end up creating that file instead.

**Comment by azturner** [comment]
@cabal95 , so would updating the FileSystem storage provider to check for NOT_A_VALIDFILESYSTEM_PATH before writing the file fix this issue then?

**Comment by cabal95** [comment]
In theory - though I think Server.MapPath() is used in a lot of places so it's possible some of these are showing up from other sources in Rock.

In our case, it was just one department that was uploading these files so in addition to switching that file type to another provider (which was on our todo list anyway) we also let them know to properly name their files other than "screenshot 11-23-2019 5:43:28pm". So we didn't really pursue it further than that.

Whether it fixes this specific symptom (meaning these files could be getting generated by other things too) I'm not sure. But it probably wouldn't hurt to put a check on that and throw an error rather than the user thinking their file uploaded correctly.

**Comment by JRichmond27** [comment]
Confirming this is still an issue in version 12.6, and that deleting _NOT_A_VALIDFILESYSTEM_PATH_ does indeed fix the problem temporarily. Thanks!

I would love to see an option in Rock that we can enable to have all file names automatically sanitized on upload. Something that replaces all whitespace with dashes, strips out all other special characters, and just for my own OCD tendancies, converts everything to lower case.

**Comment by ethan-sparkdevnetwork** [comment]
@JRichmond27 Not a bad idea. You can add it to the ideas list here: https://community.rockrms.com/Ideas

**Comment by JRichmond27** [comment]
Idea(s) added: https://community.rockrms.com/ideas/1674/file-type-improvements

If there are any other improvements you think I should have included, add them to the idea comments.

**Comment by smross** [comment]
So, we're still experiencing this problem in production on 14.3.1

I'm not sure _how_ the file named "NOT_A_VALIDFILESYSTEM_PATH" is getting created in the wwwroot, but when it happens it breaks the merge field functionality as this issue mentions. The most recent occurrence _may_ be someone uploading a vcard file, but I haven't figured out where it came from yet.

Would it be better for me to create a new issue? I also plan on testing this scenario in Alpha

**Comment by JRichmond27** [comment]
I just confirmed that this is also still happening in 14.2.

**Comment by zackdutra** [comment]
This is happening for us in 15.2. @leahjennings @smross  @JRichmond27 @ethan-sparkdevnetwork and @cabal95 do any of you have an update for what a final fix is here?

**Comment by JRichmond27** [comment]
It's happened a couple more times for us, and every time the fix is to delete the file named "NOT_A_VALIDFILESYSTEM_PATH" from Rock's root directory on the server.

**Comment by chead4** [comment]
Thank you all for the feedback.  I will be reopening this issue.

**Comment by nairdo** [comment]
It appears that the "fix" to change the `relaxedUrlToFileSystemMapping` from true to false did not get applied to the v14.0 package.  @smross, @JRichmond27, and @zackdutra  Can you confirm that your `web.config` file still has that setting set to "true":

`<httpRuntime ... relaxedUrlToFileSystemMapping="true"`

**Comment by JRichmond27** [comment]
Ours is still set to true. Should I manually change it to false?

**Comment by nairdo** [comment]
@JRichmond27 Only *if you're comfortable doing such things* -- you could, but that's just a work-around until we get to a real fix in place.  If you end up doing that, just let us know how it goes.

**Comment by JRichmond27** [comment]
I'm definitely comfortable with editing web.config files... just wanted to make sure that's all there is to it, or if that could cause other problems with Rock. 

I'll go ahead and make the change, and then post back here if we end up running into this issue again. Thanks!

**Comment by zackdutra** [comment]
@nairdo ours is set to true

**Comment by smross** [comment]
@nairdo our Production environment and our primary Test environment (used for Alpha & Beta) is set to `true`

**Comment by chead4** [comment]
@zackdutra Hi Zack, appreciate the update.  We're still digging into why this issue keeps popping up.

**Comment by chead4** [comment]
Our planned fix has caused a different issue that was uncovered during v16.3 alpha testing, so we are going to have to find a new solution to this bug.

**Comment by nairdo** [comment]
For anyone who's watching this issue and trying to understand the relationship of these fixes... The original fix (https://github.com/SparkDevNetwork/Rock/commit/7bd98a605c85ae84a12db21fa5541eb97a5d2ea1) caused an issue discovered during v16.3 alpha testing.  In their situation, a Site.Master file was attempting to reference/map a file with an 'illegal' character in the name.  So, the recent fix (https://github.com/SparkDevNetwork/Rock/commit/6dc7dcabcede5d4c6634aeb174335fcf1190152b) addresses that situation and we are going to resume the original fix to change the web.config's `httpRuntime` setting back to `relaxedUrlToFileSystemMapping="false"` in the v16.4 version.

**Comment by JimMichael** [comment]
We're getting this exception on (seemingly) every request to our site. We're still on 16.2 at this point... is this happening because we don't yet have the fix above for 16.4?

![image](https://github.com/SparkDevNetwork/Rock/assets/3903408/e8eedbdd-2e9c-4a98-b69f-6504af3da47f)

**Comment by chead4** [comment]
@JimMichael Hi Jim The issue has been fixed in v16.4 so you will no longer receive these exceptions.  You can delete the  NOT_A_VALID_FILESYSTEM_PATH file manually in v16.2.  Just keep in mind that if the file reappears, the exceptions will come back as well.
